rust-mode.el

```diff
@@ -1292,7 +1292,17 @@ whichever comes first."
 (defvar rust-imenu-generic-expression
   (append (mapcar #'(lambda (x)
                       (list (capitalize x) (rust-re-item-def-imenu x) 1))
-                  '("async fn" "enum" "struct" "union" "type" "mod" "fn" "trait" "impl"))
+                  ;; Listed in reverse order of how they are displayed (when found).
+                  '(
+                    "fn"
+                    "async\sfn"
+                    "pub\sasync\sfn"
+                    "pub\sfn"
+                    "pub\s*(\s*crate\s*)\s*fn"
```

Comment:
do we really want to have, pub(fn) and fn as different categories? It seems a bt counter-intuitive to me.

Reply:
These are hard questions, since many people probably have different preferences.

That said, good software engineering suggests making a public API and an private implementation.   Having imenu separate these too seems to follow.

Reply:
Matching pub(super|crate ... seems like a good idea.

Reply:
what about pub(crate) vs pub(super) vs any other granularity level?  I would guess that the distinction between function vs enum vs trait is more meaningful that the one between the different visibility levels of fn.

Reply:
cc @brotzeit.